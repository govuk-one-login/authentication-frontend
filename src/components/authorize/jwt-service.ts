import type { JwtServiceInterface } from "./types.js";
import {
  getOrchStubToAuthSigningPublicKey,
  getOrchToAuthSigningPublicKey,
} from "../../config.js";
import { JwtClaimsValueError, JwtValidationError } from "../../utils/error.js";
import type { Claims } from "./claims-config.js";
import {
  requiredClaimsKeys,
  getKnownClaims,
  getKnownStubClaims,
} from "./claims-config.js";
import * as jose from "jose";

export class JwtService implements JwtServiceInterface {
  private readonly publicKey;
  private readonly stubPublicKey;

  constructor(
    publicKey = getOrchToAuthSigningPublicKey(),
    stubPublicKey = getOrchStubToAuthSigningPublicKey()
  ) {
    this.publicKey = publicKey;
    this.stubPublicKey = stubPublicKey;
  }

  async getPayloadWithValidation(jwt: string): Promise<Claims> {
    let claims: jose.JWTPayload;

    try {
      claims = await this.validate(jwt);
    } catch (error) {
      throw new JwtValidationError(
        error instanceof Error ? error.message : "Unknown error validating JWT"
      );
    }

    if (claims["claim"] !== undefined) {
      this.validateClaimObject(claims["claim"] as string);
    }

    return this.validateCustomClaims(claims);
  }

  async validate(jwt: string): Promise<jose.JWTPayload> {
    try {
      return await this.validateUsingKey(jwt, this.publicKey);
    } catch (error) {
      if (this.stubPublicKey === "" || this.stubPublicKey === "UNKNOWN") {
        throw error;
      }
      return await this.validateUsingKey(jwt, this.stubPublicKey);
    }
  }

  async validateUsingKey(jwt: string, key: string): Promise<jose.JWTPayload> {
    const keyObject = await jose.importSPKI(key, "ES256");

    const result = await jose.jwtVerify(jwt, keyObject, {
      requiredClaims: requiredClaimsKeys,
      clockTolerance: 30,
    });

    return result.payload;
  }

  validateCustomClaims(claims: any): Claims {
    const requiredClaims = getKnownClaims();
    const stubClaims = getKnownStubClaims();

    Object.keys(requiredClaims).forEach((claim) => {
      if (
        requiredClaims[claim] !== claims[claim] &&
        stubClaims[claim] !== claims[claim]
      ) {
        throw new JwtClaimsValueError(`${claim} has incorrect value`);
      }
    });
    return claims;
  }

  validateClaimObject(claim: string): string {
    try {
      JSON.parse(claim);
      return claim;
    } catch {
      throw new JwtValidationError("claim object is not a valid json object");
    }
  }
}
