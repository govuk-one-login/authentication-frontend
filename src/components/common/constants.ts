import { PATH_NAMES } from "../../app.constants";
import { getNextState } from "./state-machine/state-machine";

export const SECURITY_CODE_ERROR = "actionType";

export enum SecurityCodeErrorType {
  MfaMaxCodesSent = "mfaMaxCodesSent",
  MfaBlocked = "mfaBlocked",
  MfaMaxRetries = "mfaMaxRetries",
  OtpMaxCodesSent = "otpMaxCodesSent",
  OtpBlocked = "otpBlocked",
  OtpMaxRetries = "otpMaxRetries",
  EmailMaxCodesSent = "emailMaxCodesSent",
  EmailBlocked = "emailBlocked",
  EmailMaxRetries = "emailMaxRetries",
  AuthAppMfaNoCodeValidator = "authAppMfaNoCodeValidator",
  AuthAppMfaMaxRetries = "authAppMfaMaxRetries",
  ChangeSecurityCodesEmailMaxCodesSent = "changeSecurityCodesEmailMaxCodesSent",
  ChangeSecurityCodesEmailBlocked = "changeSecurityCodesEmailBlocked",
  ChangeSecurityCodesEmailMaxRetries = "changeSecurityCodesEmailMaxRetries",
  InvalidPasswordResetCodeMaxRetries = "invalidPasswordResetCodeMaxRetries",
}

export const ERROR_CODES = {
  SESSION_ID_MISSING_OR_INVALID: 1000,
  RESET_PASSWORD_LINK_MAX_RETRIES_REACHED: 1022,
  RESET_PASSWORD_LINK_BLOCKED: 1023,
  NEW_PASSWORD_SAME_AS_EXISTING: 1024,
  ENTERED_INVALID_MFA_MAX_TIMES: 1027,
  VERIFY_EMAIL_CODE_REQUEST_BLOCKED: 1029,
  VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED: 1030,
  VERIFY_PHONE_NUMBER_MAX_CODES_SENT: 1032,
  VERIFY_EMAIL_MAX_CODES_SENT: 1031,
  ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES: 1033,
  ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES: 1034,
  MFA_SMS_MAX_CODES_SENT: 1025,
  MFA_CODE_REQUESTS_BLOCKED: 1026,
  INVALID_MFA_CODE: 1035,
  INVALID_VERIFY_EMAIL_CODE: 1036,
  INVALID_VERIFY_PHONE_NUMBER_CODE: 1037,
  ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES: 1039,
  INVALID_PASSWORD_MAX_ATTEMPTS_REACHED: 1028,
  RESET_PASSWORD_INVALID_CODE: 1021,
  PASSWORD_IS_COMMON: 1040,
  AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED: 1042,
  AUTH_APP_INVALID_CODE: 1043,
  ACCOUNT_LOCKED: 1045,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT: 1046,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED: 1047,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE: 1048,
};

export const ERROR_CODE_MAPPING: { [p: string]: string } = {
  [ERROR_CODES.ACCOUNT_LOCKED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.INVALID_PASSWORD_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.MFA_SMS_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxCodesSent
  ),
  [ERROR_CODES.MFA_CODE_REQUESTS_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_MFA_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpMaxCodesSent
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.OtpMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxRetries
  ),
  [ERROR_CODES.ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.InvalidPasswordResetCodeMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxCodesSent
  ),
  [ERROR_CODES.AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.AuthAppMfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxCodesSent
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxRetries
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_WAIT"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailBlocked
    ),
};

export function getErrorPathByCode(errorCode: number): string | undefined {
  const nextPath = ERROR_CODE_MAPPING[errorCode.toString()];

  if (!nextPath) {
    return undefined;
  }

  return nextPath;
}

export function pathWithQueryParam(
  path: string,
  queryParam?: string,
  value?: string | SecurityCodeErrorType
): string {
  if (queryParam && value) {
    const queryParams = new URLSearchParams({
      [queryParam]: value,
    }).toString();

    return path + "?" + queryParams;
  }

  return path;
}

export function getNextPathAndUpdateJourney(
  req: any,
  path: string,
  event: string,
  ctx?: any,
  sessionId?: string
): string {
  const nextState = getNextState(path, event, ctx);

  req.session.user.journey = {
    nextPath: nextState.value,
    optionalPaths:
      Object.keys(nextState.meta).length > 0
        ? nextState.meta["AUTH." + nextState.value].optionalPaths
        : [],
  };

  req.log.info(
    `User journey transitioned from ${req.path} to ${nextState.value} with session id ${sessionId}`
  );

  if (!nextState) {
    throw Error(
      `Invalid user journey. No transition found from ${path} with event ${event}`
    );
  }

  return nextState.value;
}

export const JOURNEY_TYPE = {
  SIGN_IN: "sign-in",
  CREATE_ACCOUNT: "create-account",
};
