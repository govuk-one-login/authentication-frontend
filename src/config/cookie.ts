import type { CookieOptions } from "csurf";

export function getCSRFCookieOptions(isProdEnv: boolean): CookieOptions {
  return {
    httpOnly: isProdEnv,
    secure: isProdEnv,
  };
}
