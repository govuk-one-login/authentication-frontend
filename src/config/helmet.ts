import helmet from "helmet";
import { Request, Response } from "express";
// Helmet does not export the config type - This is the way the recommend getting it on GitHub.
export const helmetConfiguration: Parameters<typeof helmet>[0] = {
  contentSecurityPolicy: {
    directives: {
      defaultSrc: ["'self'"],
      styleSrc: ["'self'"],
      scriptSrc: [
        "'self'",
        // eslint-disable-next-line @typescript-eslint/no-unnecessary-type-assertion
        (req: Request, res: Response): string =>
          `'nonce-${res.locals.scriptNonce}'`,
        "'sha256-+6WnXIl4mbFTCARd8N3COQmT3bJJmo32N8q8ZSQAIcU='",
        "https://www.googletagmanager.com",
        "https://www.google-analytics.com",
        "https://ssl.google-analytics.com",
      ],
      imgSrc: [
        "'self'",
        "data:",
        "https://www.googletagmanager.com",
        "https://www.google-analytics.com",
      ],
      objectSrc: ["'none'"],
      connectSrc: ["'self'", "https://www.google-analytics.com"],
    },
  },
  dnsPrefetchControl: {
    allow: false,
  },
  frameguard: {
    action: "deny",
  },
  hsts: {
    maxAge: 31536000, // 1 Year
    preload: true,
    includeSubDomains: true,
  },
  referrerPolicy: false,
  permittedCrossDomainPolicies: false,
  expectCt: false,
};
