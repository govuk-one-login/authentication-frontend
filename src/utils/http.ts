import type {
  AxiosInstance,
  AxiosRequestConfig,
  AxiosError,
  AxiosResponse,
  AxiosRequestHeaders,
  CreateAxiosDefaults,
} from "axios";
import axios from "axios";
import {
  getApiKey,
  getFrontendApiBaseUrl,
  supportHttpKeepAlive,
} from "../config.js";
import type { ApiResponseResult } from "../types.js";
import { HTTP_STATUS_CODES } from "../app.constants.js";
import { ApiError } from "./error.js";
import type { Request } from "express";
import { createPersonalDataHeaders } from "@govuk-one-login/frontend-passthrough-headers";
import { logger } from "./logger.js";
import { Agent } from "https";

type CustomAxiosRequestHeaders = Partial<AxiosRequestHeaders>;

const headers: CustomAxiosRequestHeaders = {
  Accept: "application/json",
  "Content-Type": "application/json; charset=utf-8",
  "Access-Control-Allow-Credentials": "true",
  "X-Requested-With": "XMLHttpRequest",
};

export interface ConfigOptions {
  sessionId?: string;
  clientSessionId?: string;
  validationStatuses?: number[];
  persistentSessionId?: string;
  baseURL?: string;
  userLanguage?: string;
  reauthenticate?: boolean;
}

export function createApiResponse<T>(
  response: AxiosResponse,
  status: number[] = [HTTP_STATUS_CODES.OK]
): ApiResponseResult<T> {
  return {
    success: status.includes(response.status),
    data: response.data,
  };
}

function getSecurityHeaders(path: string, req: Request, baseUrl?: string) {
  let personalDataHeaders = {};
  let url = "";
  try {
    const basePath = baseUrl || getFrontendApiBaseUrl();
    url = new URL(path, basePath).toString();
    personalDataHeaders = createPersonalDataHeaders(url, req);
  } catch (err) {
    logger.warn(
      `Called with ${url}. Failed to set security headers due to: ${err.message}`
    );
  }
  return personalDataHeaders;
}

export function getInternalRequestConfigWithSecurityHeaders(
  options: ConfigOptions,
  req: Request,
  path: string
): AxiosRequestConfig {
  const config: AxiosRequestConfig = {
    headers: {
      "X-API-Key": getApiKey(),
      ...getSecurityHeaders(path, req, options.baseURL),
    },
    proxy: false,
  };

  if (options.sessionId) {
    config.headers["Session-Id"] = options.sessionId;
  }

  if (options.clientSessionId) {
    config.headers["Client-Session-Id"] = options.clientSessionId;
  }

  if (options.validationStatuses) {
    config.validateStatus = function (status: number) {
      return options.validationStatuses.includes(status);
    };
  }

  if (options.persistentSessionId) {
    config.headers["di-persistent-session-id"] = options.persistentSessionId;
  }

  if (options.baseURL) {
    config.baseURL = options.baseURL;
  }

  if (options.reauthenticate) {
    config.headers["Reauthenticate"] = options.reauthenticate;
  }

  if (options.userLanguage) {
    config.headers["User-Language"] = options.userLanguage;
  }

  return config;
}

export class Http {
  private instance: AxiosInstance;

  get client(): AxiosInstance {
    return this.instance || this.initHttp();
  }

  private static handleError(error: AxiosError) {
    let apiError;

    if (error.response && error.response.data) {
      apiError = new ApiError(
        error.message,
        error.response.status,
        error.response.data
      );
    } else {
      apiError = new ApiError(error.message);
    }

    return Promise.reject(apiError);
  }

  private initHttp() {
    const http = axios.create({
      baseURL: getFrontendApiBaseUrl(),
      headers: headers,
      validateStatus: (status) => {
        return (
          status >= HTTP_STATUS_CODES.OK &&
          status <= HTTP_STATUS_CODES.BAD_REQUEST
        );
      },
      ...getAdditionalAxiosConfig(),
    });

    http.interceptors.response.use(
      (response) => response,
      (error) => Http.handleError(error)
    );

    this.instance = http;
    return http;
  }
}

export function getAdditionalAxiosConfig(): CreateAxiosDefaults {
  return supportHttpKeepAlive()
    ? { httpsAgent: new Agent({ keepAlive: true }) }
    : {};
}

export const http = new Http();
