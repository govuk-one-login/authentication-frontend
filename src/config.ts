import { RedisConfig } from "./utils/types";
import ssm from "./utils/ssm";
import { Parameter } from "aws-sdk/clients/ssm";

export function getLogLevel(): string {
  return process.env.LOGS_LEVEL || "debug";
}

export function getApiBaseUrl(): string {
  return process.env.API_BASE_URL;
}

export function getFrontendApiBaseUrl(): string {
  return process.env.FRONTEND_API_BASE_URL;
}

export function getNodeEnv(): string {
  return process.env.NODE_ENV || "development";
}

export function getAppEnv(): string {
  return process.env.APP_ENV || "local";
}

export function getSessionExpiry(): number {
  return Number(process.env.SESSION_EXPIRY);
}

export function getSessionSecret(): string {
  return process.env.SESSION_SECRET;
}

export function getApiKey(): string {
  return process.env.API_KEY;
}

export function getGtmId(): string {
  return process.env.GTM_ID;
}

export function getRedisHost(): string {
  return process.env.REDIS_HOST ?? "redis";
}

export function getRedisPort(): number {
  return Number(process.env.REDIS_PORT) ?? 6379;
}

export function supportInternationalNumbers(): boolean {
  return process.env.SUPPORT_INTERNATIONAL_NUMBERS === "1";
}

export function supportLanguageCY(): boolean {
  return process.env.SUPPORT_LANGUAGE_CY === "1";
}

export function supportAccountRecovery(): boolean {
  return process.env.SUPPORT_ACCOUNT_RECOVERY === "1";
}

export function supportAuthOrchSplit(): boolean {
  return process.env.SUPPORT_AUTH_ORCH_SPLIT === "1";
}

export function supportAuthorizeController(): boolean {
  return process.env.SUPPORT_AUTHORIZE_CONTROLLER === "1";
}

export function getSupportLinkUrl(): string {
  return process.env.URL_FOR_SUPPORT_LINKS || "/contact-us";
}

export async function getRedisConfig(appEnv: string): Promise<RedisConfig> {
  const hostKey = `${appEnv}-${process.env.REDIS_KEY}-redis-master-host`;
  const portKey = `${appEnv}-${process.env.REDIS_KEY}-redis-port`;
  const passwordKey = `${appEnv}-${process.env.REDIS_KEY}-redis-password`;

  const params = {
    Names: [hostKey, portKey, passwordKey],
    WithDecryption: true,
  };

  const result = await ssm.getParameters(params).promise();

  if (result.InvalidParameters && result.InvalidParameters.length > 0) {
    throw Error("Invalid SSM config values for redis");
  }

  return {
    password: result.Parameters.find((p: Parameter) => p.Name === passwordKey)
      .Value,
    host: result.Parameters.find((p: Parameter) => p.Name === hostKey).Value,
    port: Number(
      result.Parameters.find((p: Parameter) => p.Name === portKey).Value
    ),
    isLocal: false,
  };
}

export function getAwsRegion(): string {
  return process.env.AWS_REGION || "eu-west-2";
}

export function getKmsKeyId(): string {
  return process.env.ENCRYPTION_KEY_ID;
}

export function getOrchToAuthSigningPublicKey(): string {
  return process.env.ORCH_TO_AUTH_SIGNING_KEY;
}

export function getOrchToAuthExpectedClientId(): string {
  return process.env.ORCH_TO_AUTH_CLIENT_ID || "UNKNOWN";
}

export function getOrchToAuthExpectedAudience(): string {
  return process.env.ORCH_TO_AUTH_AUDIENCE || "UNKNOWN";
}

export function getAccountManagementUrl(): string {
  return process.env.ACCOUNT_MANAGEMENT_URL || "http://localhost:6001";
}

export function getAnalyticsCookieDomain(): string {
  return process.env.ANALYTICS_COOKIE_DOMAIN;
}

export function getServiceDomain(): string {
  return process.env.SERVICE_DOMAIN || "localhost";
}

export function getSmartAgentApiKey(): string {
  return process.env.SMARTAGENT_API_KEY || "";
}
export function getSmartAgentApiUrl(): string {
  return process.env.SMARTAGENT_API_URL || "";
}

export function getSmartAgentWebformId(): string {
  return process.env.SMARTAGENT_WEBFORM_ID || "";
}

export function getServiceSignInLink(): string {
  return process.env.SERVICE_SIGN_IN_LINK || "https://www.gov.uk/sign-in";
}

export function getCodeRequestBlockDurationInMinutes(): number {
  return Number(process.env.CODE_REQUEST_BLOCKED_MINUTES) || 15;
}

export function getCodeEnteredWrongBlockDurationInMinutes(): number {
  return Number(process.env.CODE_ENTERED_WRONG_BLOCKED_MINUTES) || 15;
}

export function getAccountRecoveryCodeEnteredWrongBlockDurationInMinutes(): number {
  return (
    Number(process.env.ACCOUNT_RECOVERY_CODE_ENTERED_WRONG_BLOCKED_MINUTES) ||
    15
  );
}

export function getPasswordResetCodeEnteredWrongBlockDurationInMinutes(): number {
  return (
    Number(process.env.PASSWORD_RESET_CODE_ENTERED_WRONG_BLOCKED_MINUTES) || 15
  );
}

export function supportFrameAncestorsFormActionsCspHeaders(): boolean {
  return process.env.FRAME_ANCESTORS_FORM_ACTIONS_CSP_HEADERS === "1";
}

export function support2FABeforePasswordReset(): boolean {
  return process.env.SUPPORT_2FA_B4_PASSWORD_RESET === "1";
}

export function supportAccountInterventions(): boolean {
  return process.env.SUPPORT_ACCOUNT_INTERVENTIONS === "1";
}

export function supportReauthentication(): boolean {
  return process.env.SUPPORT_REAUTHENTICATION === "1";
}
