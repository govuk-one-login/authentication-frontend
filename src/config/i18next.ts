import { LOCALE } from "../app.constants";
import { getServiceDomain } from "../config";

export function i18nextConfigurationOptions(
  path: string
): Record<string, unknown> {
  return {
    debug: false,
    fallbackLng: LOCALE.EN,
    preload: [LOCALE.EN],
    supportedLngs: [LOCALE.EN, LOCALE.CY],
    backend: {
      loadPath: path,
      allowMultiLoading: true,
    },
    detection: {
      lookupCookie: "lng",
      lookupQuerystring: "lng",
      order: ["querystring", "cookie"],
      caches: ["cookie"],
      ignoreCase: true,
      cookieSecure: true,
      cookieDomain: getServiceDomain(),
      cookieSameSite: "",
    },
  };
}
