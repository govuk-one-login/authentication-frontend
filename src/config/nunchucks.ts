import express from "express";
import * as nunjucks from "nunjucks";
import i18next from "i18next";
import { Environment } from "nunjucks";

export function configureNunjucks(
  app: express.Application,
  viewsPath: string[]
): Environment {
  const nunjucksEnv: nunjucks.Environment = nunjucks.configure(viewsPath, {
    autoescape: true,
    express: app,
    noCache: true,
  });

  nunjucksEnv.addFilter("translate", function (key: string, options?: any) {
    const translate = i18next.getFixedT(this.ctx.i18n.language);
    return translate(key, options);
  });

  return nunjucksEnv;
}
