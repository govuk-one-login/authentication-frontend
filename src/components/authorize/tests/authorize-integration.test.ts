import { describe } from "mocha";
import { sinon, request } from "../../../../test/utils/test-utils.js";
import nock from "nock";
import decache from "decache";
import { HTTP_STATUS_CODES, PATH_NAMES } from "../../../app.constants.js";
import type {
  AuthorizeServiceInterface,
  JwtServiceInterface,
  KmsDecryptionServiceInterface,
  StartAuthResponse,
} from "../types.js";
import { createApiResponse } from "../../../utils/http.js";
import type { AxiosResponse } from "axios";
import {
  createJwt,
  createMockClaims,
  getPrivateKey,
  getPublicKey,
} from "./test-data.js";
import { JwtService } from "../jwt-service.js";
import { getOrchToAuthExpectedClientId } from "../../../config.js";
describe("Integration:: authorize", () => {
  let app: any;
  let userCookieConsent = false;

  beforeEach(async () => {
    process.env.SUPPORT_AUTHORIZE_CONTROLLER = "1";
    decache("../../../app");
    decache("../authorize-service");
    decache("../kms-decryption-service");
    decache("../jwt-service");
    const authorizeService = await import("../authorize-service.js");
    const KmsDecryptionService = await import("../kms-decryption-service.js");
    const jwtService = await import("../jwt-service.js");
    const publicKey = getPublicKey();
    const privateKey = await getPrivateKey();
    const jwt = await createJwt(createMockClaims(), privateKey);

    sinon
      .stub(authorizeService, "authorizeService")
      .callsFake((): AuthorizeServiceInterface => {
        async function start() {
          return createApiResponse<StartAuthResponse>(
            fakeAxiosStartResponse(userCookieConsent)
          );
        }

        return { start };
      });

    sinon
      .stub(KmsDecryptionService, "KmsDecryptionService")
      .callsFake((): KmsDecryptionServiceInterface => {
        async function decrypt() {
          return Promise.resolve(jwt);
        }
        return { decrypt };
      });

    sinon.stub(jwtService, "JwtService").callsFake((): JwtServiceInterface => {
      return new JwtService(publicKey);
    });

    app = await (await import("../../../app.js")).createApp();
  });

  beforeEach(() => {
    nock.cleanAll();
  });

  after(() => {
    sinon.restore();
    app = undefined;
  });

  it("should redirect to /sign-in-or-create", async () => {
    await request(app, (test) =>
      test
        .get(PATH_NAMES.AUTHORIZE)
        .query({
          client_id: getOrchToAuthExpectedClientId(),
          response_type: "code",
          request: "SomeJWE",
        })
        .expect("Location", PATH_NAMES.SIGN_IN_OR_CREATE)
        .expect(302)
    );
  });

  it("should redirect to /sign-in-or-create if user cookie consent is true", async () => {
    userCookieConsent = true;
    await request(app, (test) =>
      test
        .get(PATH_NAMES.AUTHORIZE)
        .query({
          client_id: getOrchToAuthExpectedClientId(),
          response_type: "code",
          request: "SomeJWE",
        })
        .expect("Location", PATH_NAMES.SIGN_IN_OR_CREATE)
        .expect(302)
    );
  });

  it("should redirect to /sign-in-or-create with Google Analytics tag if 'result' query exists", async () => {
    await request(app, (test) =>
      test
        .get(PATH_NAMES.AUTHORIZE)
        .query({
          client_id: getOrchToAuthExpectedClientId(),
          response_type: "code",
          request: "SomeJWE",
          result: "test-result",
        })
        .expect(
          "Location",
          PATH_NAMES.SIGN_IN_OR_CREATE + "?result=test-result"
        )
        .expect(302)
    );
  });

  function fakeAxiosStartResponse(userCookieConsent: boolean): AxiosResponse {
    return {
      data: {
        client: {
          serviceType: "MANDATORY",
          clientName: "test-client",
          cookieConsentEnabled: true,
          redirectUri: "http://test-redirect.gov.uk/callback",
          state: "jasldasl12312",
          isOneLoginService: false,
        },
        user: {
          upliftRequired: false,
          identityRequired: false,
          authenticated: false,
          cookieConsent: userCookieConsent ? "accept" : "reject",
        },
      },
      status: HTTP_STATUS_CODES.OK,
    } as AxiosResponse;
  }
});
