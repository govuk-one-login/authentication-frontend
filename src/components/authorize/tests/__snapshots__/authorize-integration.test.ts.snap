// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration:: authorize should redirect to /sign-in-or-create with Google Analytics tag if 'result' query exists 1`] = `
Object {
  "taxonomyLevel1": "accounts",
  "taxonomyLevel2": "create account",
}
`;
